### PR TITLE
feature: audio player

### PR DESCRIPTION
**Pull request by moiskillnadne**
Screenshot 2023-07-20 at 17 53 47
Screenshot 2023-07-20 at 18 28 39

### CODE REVIEW

**Review by BamMironov — CHANGES_REQUESTED** [review]

File: src/entities/activity/ui/items/AudioPlayerItem.tsx [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import Box from "@mui/material/Box"
+
+import { AudioPlayerItem as AudioPlayerItemType } from "../../lib/types/item"
+
+import { AudioPlayerItemBase } from "~/shared/ui"
+
+type Props = {
+  item: AudioPlayerItemType
```
[the remark this answers is not in the document]
Comment (moiskillnadne) [reply src/entities/activity/ui/items/AudioPlayerItem.tsx:8]: Yes, I will change it in the next MR for all items

File: src/shared/ui/Items/AudioPlayer/lib/useAudioVolume.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  const unmute = () => {
+    if (audioRef.current?.volume === undefined) {
+      return
+    }
+
+    setIsMuted(false)
+    audioRef.current.volume = volume
+  }
+
+  const onVolumeChange = (volume: number) => {
```
[the remark this answers is not in the document]
Comment (moiskillnadne) [reply src/shared/ui/Items/AudioPlayer/lib/useAudioVolume.ts:49]: Done

File: src/shared/ui/Items/AudioPlayer/lib/useAudioControls.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+  const pause = () => {
+    setIsPlaying(false)
+    return audioRef.current?.pause()
+  }
+
+  return {
+    isPlaying,
+    play,
+    pause,
```
[the remark this answers is not in the document]
Comment (moiskillnadne) [reply src/shared/ui/Items/AudioPlayer/lib/useAudioControls.ts:29]: Not agree, I implemented two different methods on purpose. I did it for easy control of each event.

File: src/shared/ui/Items/AudioPlayer/ui/AudioPlayerControls.tsx [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    sm?: boolean
+  }
+}
+
+export const AudioPlayerControls = ({ isPlaying, onPlayClick, onPauseClick, isDisabled, mediaQuery }: Props) => {
+  const iconSize = mediaQuery?.sm ? "small" : "medium"
+
+  return (
+    <Box>
+      <IconButton onClick={isPlaying ? onPauseClick : onPlayClick} disabled={isDisabled}>
```
[the remark this answers is not in the document]
Comment (moiskillnadne) [reply src/shared/ui/Items/AudioPlayer/ui/AudioPlayerControls.tsx:23]: Done

File: src/shared/ui/Items/AudioPlayer/lib/useAudioVolume.ts [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+type Props = {
+  audioRef: RefObject<HTMLAudioElement>
+}
+
+type Return = {
+  isMuted: boolean
+  volume: number
+
+  mute: () => void
+  unmute: () => void
```
[the remark this answers is not in the document]
Comment (moiskillnadne) [reply src/shared/ui/Items/AudioPlayer/lib/useAudioVolume.ts:12]: The same approach I used here. I split these methods for easy use and control depending on situation

File: src/shared/ui/Items/AudioPlayer/ui/AudioPlayerControls.tsx [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+type Props = {
+  isPlaying: boolean
+  onPlayClick: () => void
+  onPauseClick: () => void
+
+  isDisabled?: boolean
+
+  mediaQuery?: {
+    sm?: boolean
+  }
```
[the remark this answers is not in the document]
Comment (moiskillnadne) [reply src/shared/ui/Items/AudioPlayer/ui/AudioPlayerControls.tsx:15]: Done

File: src/shared/ui/Items/AudioPlayer/ui/AudioPlayerDuration.tsx [hunk 7]
```diff
@@ -0,0 +1,25 @@
+import { Typography } from "@mui/material"
+import Box from "@mui/material/Box"
+
+type Props = {
+  currentDuration: string
+  totalDuration: string
+
+  mediaQuery?: {
+    sm?: boolean
+  }
```
[the remark this answers is not in the document]
Comment (moiskillnadne) [reply src/shared/ui/Items/AudioPlayer/ui/AudioPlayerDuration.tsx:10]: Done

File: src/shared/ui/Items/AudioPlayer/ui/AudioPlayerVolume.tsx [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+  value?: number
+
+  onHandleMute: () => void
+  onHandleUnmute: () => void
+
+  onChange: (volume: number) => void
+
+  mediaQuery?: {
+    sm?: boolean
+  }
```
[the remark this answers is not in the document]
Comment (moiskillnadne) [reply src/shared/ui/Items/AudioPlayer/ui/AudioPlayerVolume.tsx:19]: Done

File: src/shared/ui/Items/AudioPlayer/ui/AudipPlayerProgressBar.tsx [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+  progress: number
+  buffer?: number
+
+  isDisabled?: boolean
+
+  onProgressBarClick: (progress: number) => void
+
+  mediaQuery?: {
+    sm?: boolean
+  }
```
[the remark this answers is not in the document]
Comment (moiskillnadne) [reply src/shared/ui/Items/AudioPlayer/ui/AudipPlayerProgressBar.tsx:16]: Done

**Review by BamMironov — APPROVED** [review]

**Pull request merged**